docs/source/how-to/custom-formats.md

```diff
@@ -13,18 +13,156 @@ assumptions about the array.)
+# Facility.name: APS
+# Facility.energy: 7.00 GeV
+# Facility.xray_source: APS Undulator A
+# Scan.start_time: 2001-06-26T22:27:31
+# Detector.I0: 10cm  N2
+# Detector.I1: 10cm  N2
+# Sample.name: Cu
+# Sample.prep: Cu metal foil
+# GSE.EXTRA:  config 1
+# ///
+# Cu foil Room Temperature
+# measured at beamline 13-ID
+#----
+# energy i0 itrans mutrans
+  8779.0  149013.7  550643.089065  -1.3070486
+  8789.0  144864.7  531876.119084  -1.3006104
+  8799.0  132978.7  489591.10592  -1.3033816
+  8809.0  125444.7  463051.104096  -1.3059724
+  8819.0  121324.7  449969.103983  -1.3107085
+  8829.0  119447.7  444386.117562  -1.3138152
+  8839.0  119100.7  440176.091039  -1.3072055
+  8849.0  117707.7  440448.106567  -1.3195882
+  8859.0  117754.7  442302.10637  -1.3233895
+  8869.0  117428.7  441944.116528  -1.3253521
+  8879.0  117383.7  442810.120466  -1.327693
+  8889.0  117185.7  443658.11566  -1.3312944
+```
+As you can see , the file's contents comprise a single table
 and a header of key--value pairs. Therefore, it makes sense to export `dataframe`
 structures and their associated metadata into this format.
 
-Take the following simple server configuration:
+Consider a folder `data` containing a number of xdi files that we wish to serve using tiled.
+An example folder containing a single `example.xdi` file can be generated by running `python -m tiled.examples.xdi`.
+```{note}
+All code below is in `tiled/examples/xdi.py`.
+The code is reproduced for reference but running the examples only requires modifying `config.yaml`.
+```
+
+# Define an Importer
+
+The first step in exposing this data via tiled is to write a function which can read xdi formatted files.
+Note this function is provided as an example and does not do the validation that would be expected from a full parser.
+
+```py
+# tiled/examples/xdi.py
+
+def read_xdi(path):
+    with open(path, "r") as f:
+        metadata = {}
+        fields = collections.defaultdict(dict)
+
+        line = f.readline()
+        m = re.match("# XDI/(\S*)\s*(\S*)?", line)
+        if not m:
+            raise AssertionError(f"not an XDI file, no XDI versioning information in first line\n{line}")
+
+        metadata["xdi_version"] = m[1]
+        metadata["extra_version"] = m[2]
+
+        field_end_re = re.compile("#\s*/{3,}")
+        header_end_re = re.compile("#\s*-{3,}")
+
+        has_comments = False
+
+        # read header
+        for line in f:
+            if line[0] != "#":
+                raise AssertionError(f"reached invalid line in header\n{line}")
+            if re.match(field_end_re, line):
+                has_comments = True
+                break
+            elif re.match(header_end_re, line):
+                break
+
+            try:
+                key, val = line[1:].strip().split(":", 1)
+                val = val.strip()
+                namespace, tag = key.split(".")
+                #TODO coerce to lower case?
+            except:
```

Comment:
Avoid bare `except:` --- this would catch `KeyboardInterrupt` for example as well. I suggest:

```py
except ValueError as err:
    raise ValueError(f"error processing line\n{line}") from err
```

That will show the specific error at the bottom but also include the original error above as its direct cause.